Add `BinaryClassifierPrecisionEfficacy` metric

R-Palazzo opened the pull request:
CU-86b3m9u4d
Resolve #711

sdv-team commented:
Task linked: [CU-86b3m9u4d SDMetrics - Add `BinaryClassifierPrecisionEfficacy` metric #711](https://app.clickup.com/t/86b3m9u4d)

amontanez24 commented on sdmetrics/single_table/data_augmentation/utils.py:
```diff
@@ -0,0 +1,126 @@
+    if classifier is not None and not isinstance(classifier, str):
+        raise TypeError('`classifier` must be a string or None.')
+
+    if classifier != 'XGBoost':
```
For these types of validations I think it's good to split them into separate smaller functions. The reason for this is that if we decide to expand this class to support more classifiers or create a new class to do that, we might want to overwrite just this part of the validation and keep everything else.

amontanez24 commented on sdmetrics/single_table/data_augmentation/utils.py:
```diff
@@ -0,0 +1,126 @@
+    if not synthetic_labels.issubset(real_labels):
+        raise ValueError(
+            f'The ``{prediction_column_name}`` column must have the same values in the real '
+            'and synthetic data. The synthetic data has the following unseen values: '
```
```suggestion
            'and synthetic data. The following values are present in the synthetic data and not the real data: '
```

amontanez24 commented on sdmetrics/single_table/data_augmentation/base.py:
```diff
@@ -0,0 +1,258 @@
+        return scores
+
+    @classmethod
+    def compute_breakdown(
```
why is this a class method?

R-Palazzo replied:
It's a class method by definition:
https://github.com/sdv-dev/SDMetrics/blob/280a93e034a33d050f47ddf576b4efb63505f1e8/sdmetrics/single_table/base.py#L168

amontanez24 commented on sdmetrics/single_table/data_augmentation/base.py:
```diff
@@ -0,0 +1,258 @@
+                The tables to transform.
+        """
+        tables_result = {}
+        for table_name, table in tables.items():
```
Isn't this a single table metric?

amontanez24 commented on sdmetrics/single_table/data_augmentation/base.py:
```diff
@@ -0,0 +1,258 @@
+
+        self._fit_preprocess(data, metadata)
+
+    def _transform_preprocess(self, tables):
```
minor: we could just call transform

R-Palazzo replied:
I find `_fit_preprocess` and `_transform_preprocess` clearer, and there is no confusion with the `fit` method of the ML model. The ML model does not have a transform method, but since we keep `_fit_preprocess`, it's consistent to name `_transform_preprocess` as well.

amontanez24 replied:
The ML model was separated out though. I think _fit_preprocess confuses me as it makes me wonder if we are fitting on preprocessed data or fitting as part of the preprocess step

amontanez24 replied:
@R-Palazzo

R-Palazzo replied:
@amontanez24 sorry I missed this, done in [fbc5d37](https://github.com/sdv-dev/SDMetrics/pull/714/commits/fbc5d372879c7313dc46a13543158e6a629a8110)

amontanez24 commented on sdmetrics/single_table/data_augmentation/base.py:
```diff
@@ -0,0 +1,258 @@
+        target_probabilities = self._classifier.predict_proba(train_data)[:, 1]
+        precision, recall, thresholds = precision_recall_curve(train_target, target_probabilities)
+        # To assess the preicision efficacy, we have to fix the recall and reciprocally
+        metric = precision if self.metric_name == 'recall' else recall
```
can't you just make this self._metric_to_fix?

amontanez24 commented on sdmetrics/single_table/data_augmentation/base.py:
```diff
@@ -0,0 +1,258 @@
+            fixed_recall_value,
+        )
+        result = {}
+        metric = cls()
+        metric._fit(
```
something about this seems strang4e. Maybe we should make a separate class to represent the object that fits/trains the classifier and makes predictions. Then this metric can just have a reference to that object

R-Palazzo commented:
Thanks for the review @amontanez24. I addressed the comments in [791c7d5](https://github.com/sdv-dev/SDMetrics/pull/714/commits/791c7d5599545c923ac24c94518fdf3bf0ec09da) and [aafdd1e](https://github.com/sdv-dev/SDMetrics/pull/714/commits/aafdd1e1abc48732abb92c475585fe02af1dcd79) with some refactoring as you recommended

R-Palazzo commented on sdmetrics/visualization.py:
```diff
@@ -32,7 +32,7 @@ def wrapper(*args, **kwargs):
             if 'ZMQInteractiveShell' in ipython_interpreter and 'iframe' in renderers:
                 # This means we are using jupyter notebook
-                pio.renderers.default = 'vscode'
+                pio.renderers.default = 'iframe'
```
I found that this change was accidentally added in a previous PR, so I reverted it.

amontanez24 reviewed (approved):
LGTM! @frances-h Can you take a look? You know the details of the algorithm better than I do
